hotstuff/src/proposer.rs

```diff
@@ -0,0 +1,139 @@
+use crate::consensus::{ConsensusMessage, Round};
+use crate::messages::{Block, QC, TC};
+use bytes::Bytes;
+use config::{Committee, Stake};
+use crypto::{PublicKey, SignatureService};
+use futures::stream::futures_unordered::FuturesUnordered;
+use futures::stream::StreamExt as _;
+use log::{debug, info};
+use network::{CancelHandler, ReliableSender};
+use primary::Certificate;
+use tokio::sync::mpsc::{Receiver, Sender};
+use tokio::time::{sleep, Duration};
+
+#[derive(Debug)]
+pub struct ProposerMessage(pub Round, pub QC, pub Option<TC>);
+
+pub struct Proposer {
+    name: PublicKey,
+    committee: Committee,
+    signature_service: SignatureService,
+    rx_mempool: Receiver<Certificate>,
+    rx_message: Receiver<ProposerMessage>,
+    tx_loopback: Sender<Block>,
+    buffer: Vec<Certificate>,
+    network: ReliableSender,
+}
+
+impl Proposer {
+    pub fn spawn(
+        name: PublicKey,
+        committee: Committee,
+        signature_service: SignatureService,
+        rx_mempool: Receiver<Certificate>,
+        rx_message: Receiver<ProposerMessage>,
+        tx_loopback: Sender<Block>,
+    ) {
+        tokio::spawn(async move {
+            Self {
+                name,
+                committee,
+                signature_service,
+                rx_mempool,
+                rx_message,
+                tx_loopback,
+                buffer: Vec::new(),
+                network: ReliableSender::new(),
+            }
+            .run()
+            .await;
+        });
+    }
+
+    /// Helper function. It waits for a future to complete and then delivers a value.
+    async fn waiter(wait_for: CancelHandler, deliver: Stake) -> Stake {
+        let _ = wait_for.await;
+        deliver
+    }
+
+    async fn make_block(&mut self, round: Round, qc: QC, tc: Option<TC>) {
+        // Generate a new block.
+        let block = Block::new(
+            qc,
+            tc,
+            self.name,
+            round,
+            /* payload */ self.buffer.drain(..).collect(),
```

Comment:
Is there a point to add more than one cert? The latest should have all other in its causal history